capsules/src/net/udp/udp_recv.rs

```diff
@@ -0,0 +1,63 @@
+use core::cell::Cell;
+use net::ipv6::ip_utils::IPAddr;
+use net::ipv6::ipv6::IP6Header;
+use net::ipv6::ipv6_recv::IP6RecvClient;
+use net::udp::udp::UDPHeader;
+
+pub trait UDPRecvClient {
+    fn receive(
+        &self,
+        src_addr: IPAddr,
+        dst_addr: IPAddr,
+        src_port: u16,
+        dst_port: u16,
+        payload: &[u8],
+    );
+}
+
+pub trait UDPReceiver<'a> {
+    fn set_client(&self, client: &'a UDPRecvClient);
+}
+
+pub struct UDPRecvStruct<'a> {
```

Comment:
This was originally made a trait because it was copied from the IP layer, where it matters that it be a trait because in theory in the future we could have the IP layer sit on top of multiple link layers. I can't imagine we will have UDP on top of any other network layer than IP, so I think it is fine for UDPRecvStruct to become UDPReceiver. I'll do that